gusto/linear_solvers.py

```diff
@@ -229,6 +237,332 @@ def solve(self):
         theta.assign(self.theta)
 
 
+class HybridisedCompressibleSolver(TimesteppingSolver):
+    """
+    Timestepping linear solver object for the compressible equations
+    in theta-pi formulation with prognostic variables u, rho, and theta.
+
+    This solver follows the following strategy:
+
+    (1) Analytically eliminate theta (introduces error near topography)
+
+    (2a) Formulate the resulting mixed system for u and rho using a
+         hybridized mixed method. This breaks continuity in the
+         linear perturbations of u, and introduces a new unknown on the
+         mesh interfaces approximating the average of the Exner pressure
+         perturbations. These trace unknowns also act as Lagrange
+         multipliers enforcing normal continuity of the "broken" u variable.
+
+    (2b) Statically condense the block-sparse system into a single system
+         for the Lagrange multipliers. This is the only globally coupled
+         system requiring a linear solver.
+
+    (2c) Using the computed trace variables, we locally recover the
+         broken velocity and density perturbations. This is accomplished
+         in two stages:
+         (i): Recover rho locally using the multipliers.
+         (ii): Recover "broken" u locally using rho and the multipliers.
+
+    (2d) Project the "broken" velocity field into the HDiv-conforming
+         space using local averaging.
+
+    (3) Reconstruct theta
+
+    :arg state: a :class:`.State` object containing everything else.
+    :arg quadrature degree: tuple (q_h, q_v) where q_h is the required
+                            quadrature degree in the horizontal direction
+                            and q_v is that in the vertical direction.
+    :arg solver_parameters (optional): solver parameters for the
+                                       trace system.
+    :arg overwrite_solver_parameters: boolean, if True use only the solver_parameters
+                                      that have been passed in, if False then update.
+                                      the default solver parameters with the
+                                      solver_parameters passed in.
+    :arg moisture (optional): list of names of moisture fields.
+    """
+
+    # Solver parameters for the Lagrange multiplier system
+    # NOTE: The reduced operator is not symmetric
+    solver_parameters = {'ksp_type': 'gmres',
```

Comment:
Just a general remark: Can we do better here in terms of an iterative solver? I thought bicg-stab would be a good choice. It works for nonsymmetric systems, seemed stable enough in serial runs. However, running in parallel seems to dramatically increase the number of iterations to reduce the residual by the specified tolerance. I'm talking going from 6 iterations in serial to around 20 for 4 processes (rtol 1e-8).